File: Master 1/Python && R/Exos/pendu.py
import random

def display_hidden_word(word, lettres_trouvees):
    return "".join([letter if letter in lettres_trouvees else "_" for letter in word])

def play():

    # List of possible words
    words = [
        "python", "wesley", "ordinateur", "chocolat", "grenouille",
        "cuisine", "aventure", "papillon", "voyage",
        "soleil", "guitare", "musique", "montagne", "nuage"
        ]
    
    # Sélectioner un mot au hasard 
    secret_word = random.choice(words)  # Randomly select a word

    guessed_letters = set()
    attempted_letters = set()
    # Nombre de choix possibles
    remaining_attempts = len(secret_word) 
    print(secret_word)

    print("\n\n######################################################################### \n\n")
    print("Bienvenu dans le pendu, faites tout possible pour ne pas être pendu \n")
    print("Veuillez déviner le bon mot avant d' être pendu.")
    print("\n\n######################################################################### \n")

    while remaining_attempts > 0:
        print("\n\n##########")
        print("\nMot à déviner: ", display_hidden_word(secret_word, guessed_letters))
        print(f"Mots choisi déjà: {', '.join(sorted(attempted_letters))}")
        print(f"Nombre de choix restant: {remaining_attempts}")

        letter = input("Enter une lettre: ").lower()

        # Vérifications
        if len(letter) != 1 or not letter.isalpha():
            print("Veuillez entrer une seule lettre .")
            continue
        if letter in attempted_letters:
            print("Vous avez déjà choisi cette lettre.")
            continue

        attempted_letters.add(letter)

        if letter in secret_word:
            guessed_letters.add(letter)
            print(f"Succès '{letter}' est das le mot.")
        else:
            remaining_attempts -= 1
            print(f"Erreur, '{letter}' n' est pas dans le mot.")
        print("\n##########")

        # Check for victory
        if all(l in guessed_letters for l in secret_word):
            print("\n\n***********************************************************\n")
            print("Félicitation, Vous avez trouvé tout les lettres du mot !", secret_word)
            print("\n***********************************************************\n")
            break
    else:
        print("\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
        print("Vous avez perdu ! Le mot était:", secret_word)
        print("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")

# Run the game
if __name__ == "__main__":
    play()

File: Master 1/Python && R/Exos/pendu.R
library(stringr)

display_hidden_word <- function(word, guessed_letters) {
  str_replace_all(word, setdiff(strsplit(word, "")[[1]], guessed_letters), "_")
}

play <- function() {
  # Liste des mots
  words <- c("python", "wesley", "ordinateur", "chocolat", "grenouille",
             "cuisine", "aventure", "papillon", "voyage",
             "soleil", "guitare", "musique", "montagne", "nuage")
  
  # Selectiooner un mot
  secret_word <- sample(words, 1)
  
  guessed_letters <- character()
  attempted_letters <- character()
  remaining_attempts <- length(secret_word)
  
  print(secret_word)

  cat("\n\n######################################################################### \n\n")
  cat("Bienvenu dans le pendu, faites tout possible pour ne pas être pendu \n")
  cat("Veuillez déviner le bon mot avant d' être pendu.")
  cat("\n\n######################################################################### \n")

  while (remaining_attempts > 0) {
    cat("\n\n##########\n")
    cat("Mot à déviner: ", display_hidden_word(secret_word, guessed_letters))
    cat(paste("Mots choisi déjà: ", paste(attempted_letters, collapse = ", ")))
    cat(paste("Nombre de choix restant: ", remaining_attempts))

    letter <- readline(prompt = "Enter une lettre: ")
    
    # Check if input is valid
    if (length(letter) != 1 | !grepl("[a-zA-Z]", letter)) {
      cat("Veuillez entrer une seule lettre .\n")
      next
    }
    
    if (!(letter %in% attempted_letters)) {
      attempted_letters <- c(attempted_letters, letter)
      
      if (letter %in% secret_word) {
        guessed_letters <- c(guessed_letters, letter)
        cat(paste("Succès '", letter, "' est dans le mot.\n"))
      } else {
        remaining_attempts <- remaining_attempts - 1
        cat(paste("Erreur, '", letter, "' n'est pas dans le mot.\n"))
      }
    }
    cat("\n##########\n")

    # Check for victory
    if (all(secret_word %in% guessed_letters)) {
      cat("\n\n***********************************************************\n")
      cat(paste("Félicitation, Vous avez trouvé toutes les lettres du mot !", secret_word))
      cat("\n***********************************************************\n")
      break
    }
  }
  
  # If game ends without winning
  if (remaining_attempts == 0) {
    cat("\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
    cat(paste("Vous avez perdu ! Le mot était:", secret_word))
    cat("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
  }
}

# Run the game
play()
